File: clients/native/examples/python-examples/websocket/textsend.py
import asyncio
import json
import websockets

self_address_request = json.dumps({
    "type": "selfAddress"
})


async def send_text_without_reply():
    message = "Hello Nym!"

    uri = "ws://localhost:1977"
    async with websockets.connect(uri) as websocket:
        await websocket.send(self_address_request)
        self_address = json.loads(await websocket.recv())
        print("our address is: {}".format(self_address["address"]))

        text_send = json.dumps({
            "type": "send",
            "message": message,
            "recipient": self_address["address"],
            "withReplySurb": False,
        })

        print("sending '{}' (*without* reply SURB) over the mix network...".format(message))
        await websocket.send(text_send)

        print("waiting to receive a message from the mix network...")
        received_message = await websocket.recv()
        print("received '{}' from the mix network".format(received_message))


async def send_text_with_reply():
    message = "Hello Nym!"

    uri = "ws://localhost:1977"
    async with websockets.connect(uri) as websocket:
        await websocket.send(self_address_request)
        self_address = json.loads(await websocket.recv())
        print("our address is: {}".format(self_address["address"]))

        text_send = json.dumps({
            "type": "send",
            "message": message,
            "recipient": self_address["address"],
            "withReplySurb": True,
        })

        print("sending '{}' (*with* reply SURB) over the mix network...".format(message))
        await websocket.send(text_send)

        print("waiting to receive a message from the mix network...")
        received_message = json.loads(await websocket.recv())
        print("received '{}' from the mix network".format(received_message))

        # use the received surb to send an anonymous reply!
        reply_surb = received_message["replySurb"]

        reply_message = "hello from reply SURB!"
        reply = json.dumps({
            "type": "reply",
            "message": reply_message,
            "replySurb": reply_surb
        })

        print("sending '{}' (using reply SURB!) over the mix network...".format(reply_message))
        await websocket.send(reply)

        print("waiting to receive a message from the mix network...")
        received_message = await websocket.recv()
        print("received '{}' from the mix network".format(received_message))



# asyncio.get_event_loop().run_until_complete(send_text_without_reply())
asyncio.get_event_loop().run_until_complete(send_text_with_reply())

File: clients/native/examples/go-examples/websocket/text/textsend.go
package main

import (
	"encoding/json"
	"fmt"

	"github.com/gorilla/websocket"
)

func getSelfAddress(conn *websocket.Conn) string {
	selfAddressRequest, err := json.Marshal(map[string]string{"type": "selfAddress"})
	if err != nil {
		panic(err)
	}

	if err = conn.WriteMessage(websocket.TextMessage, []byte(selfAddressRequest)); err != nil {
		panic(err)
	}

	responseJSON := make(map[string]interface{})
	err = conn.ReadJSON(&responseJSON)
	if err != nil {
		panic(err)
	}

	return responseJSON["address"].(string)
}

func sendTextWithoutReply() {
	message := "Hello Nym!"

	uri := "ws://localhost:1977"

	conn, _, err := websocket.DefaultDialer.Dial(uri, nil)
	if err != nil {
		panic(err)
	}
	defer conn.Close()

	selfAddress := getSelfAddress(conn)
	fmt.Printf("our address is: %v\n", selfAddress)
	sendRequest, err := json.Marshal(map[string]interface{}{
		"type":          "send",
		"recipient":     selfAddress,
		"message":       message,
		"withReplySurb": false,
	})
	if err != nil {
		panic(err)
	}

	fmt.Printf("sending '%v' (*without* reply SURB) over the mix network...\n", message)
	if err = conn.WriteMessage(websocket.TextMessage, []byte(sendRequest)); err != nil {
		panic(err)
	}

	fmt.Printf("waiting to receive a message from the mix network...\n")
	_, receivedMessage, err := conn.ReadMessage()
	if err != nil {
		panic(err)
	}
	fmt.Printf("received %v from the mix network!\n", string(receivedMessage))
}

func sendTextWithReply() {
	message := "Hello Nym!"

	uri := "ws://localhost:1977"

	conn, _, err := websocket.DefaultDialer.Dial(uri, nil)
	if err != nil {
		panic(err)
	}
	defer conn.Close()

	selfAddress := getSelfAddress(conn)
	fmt.Printf("our address is: %v\n", selfAddress)
	sendRequest, err := json.Marshal(map[string]interface{}{
		"type":          "send",
		"recipient":     selfAddress,
		"message":       message,
		"withReplySurb": true,
	})
	if err != nil {
		panic(err)
	}

	fmt.Printf("sending '%v' (*with* reply SURB) over the mix network...\n", message)
	if err = conn.WriteMessage(websocket.TextMessage, []byte(sendRequest)); err != nil {
		panic(err)
	}

	fmt.Printf("waiting to receive a message from the mix network...\n")
	_, receivedMessage, err := conn.ReadMessage()
	if err != nil {
		panic(err)
	}
	fmt.Printf("received %v from the mix network!\n", string(receivedMessage))

	receivedMessageJSON := make(map[string]interface{})
	if err := json.Unmarshal(receivedMessage, &receivedMessageJSON); err != nil {
		panic(err)
	}

	// use the received surb to send an anonymous reply!
	replySurb := receivedMessageJSON["replySurb"]
	replyMessage := "hello from reply SURB!"

	reply, err := json.Marshal(map[string]interface{}{
		"type":      "reply",
		"message":   replyMessage,
		"replySurb": replySurb,
	})
	if err != nil {
		panic(err)
	}

	fmt.Printf("sending '%v' (using reply SURB) over the mix network...\n", replyMessage)
	if err = conn.WriteMessage(websocket.TextMessage, []byte(reply)); err != nil {
		panic(err)
	}

	fmt.Printf("waiting to receive a message from the mix network...\n")
	_, receivedMessage, err = conn.ReadMessage()
	if err != nil {
		panic(err)
	}
	fmt.Printf("received %v from the mix network!\n", string(receivedMessage))
}

func main() {
	// sendTextWithoutReply()
	sendTextWithReply()
}
